Add previous scene to tabBarOnPress - fix #2787

pietro909 opened the pull request:
Goals:

- pass the previous scene to the `tabBarOnPress` callback, together with its current arguments

Why?

- Implementing some custom logic (ie: reset the child navigator of the scene we are leaving) was not possible in other ways

Please see issue #2787

**Test plan (required)**

Being it an argument (which always exists) passed to a callback, the only tests I can deliver are the type checks and the current tests themselves.

kelset commented:
Couple things:
1. it would be helpful to modify the docs too to reflect the change ([here](https://reactnavigation.org/docs/navigators/tab#tabBarOnPress))
1. shouldn't jumpToIndex get this "history" param too?

Tbh it should be even something we have more in general, I mean to only add it for this specific case is sort of de-aligning the navigation methods... or is it just me?

pietro909 commented:
Thank you @kelset for the feedback:

1. docs updated in d3b5901
2. I don't think so because `jumpToIndex` is in charge of just go straight to the desired index (if it exists), so as a consumer of _react-navigation_ I wouldn't expect it to take any decision beyond the linear jump itself

> it should be even something we have more in general

I agree, can you help me to come up with a list of the possible entries to be updated? I'm quite new to this library.

Thanks!

kelset commented:
1. thanks for updating the docs
1. thinking more about it yeah, I agree. Moreover, this is in the context of somebody overriding a "default method", so it makes sense that we provide the dev with all these information.
1.  > I agree, can you help me to come up with a list of the possible entries to be updated? I'm quite new to this library.

I'm quite sure that this method is "unique" at the moment, moreover in the Stack scenario you always know the previous route (= it's linear). And in the Drawer AFAIK there's no method to override the onPress method.

---

By the way, while checking all the files to write the third point I've noticed that you haven't modified `TabBarTop` so that needs to be fixed before merging ;)

spencercarli commented:
Just noting this for myself or @kelset when we make the next release notes that this will be a breaking API change (though certainly worth it) and we should notate that.

pietro909 commented:
I've updated `TabBarTop` and rebased against `master` branch.

Though I've noticed that I get this when I run `yarn test`:

<img width="644" alt="screen shot 2017-10-25 at 08 56 47" src="https://user-images.githubusercontent.com/2094604/31984672-a930143a-b962-11e7-8586-263e434cc39e.png">

kelset commented:
Ok, thanks.
This weekend I'll review it.

kelset commented:
Hey @pietro909 sorry for not giving you feedback earlier, we have been busy with `beta.16` and the issues it spawned - now most of them should be fixed so we can focus again on other PRs.
Can you please fix the merge conflicts?

pietro909 commented:
Thank you @kelset I've fixed conflicts and rebased. Please let me know if there's something else.

I'm looking forward to having this one in the master since we use this feature at Knipster :-)

kelset commented:
LGTM - @spencercarli or @vonovak can you take a look at it too? I think we can merge.

vonovak commented on src/views/TabView/TabBarBottom.js:
```diff
@@ -187,7 +193,7 @@ class TabBarBottom extends React.PureComponent<Props> {
             const focused = index === navigation.state.index;
             const scene = { route, index, focused };
-            const onPress = getOnPress(scene);
+            const onPress = getOnPress(previousScene, scene);
```
As noted by @spencercarli this is a breaking change. I think we shouldn't introduce breaking changes unless there are very good reasons for doing so. In this case, changing `const onPress = getOnPress(previousScene, scene);` to `const onPress = getOnPress(scene, previousScene);` and changes in a couple more lines will make this a non-breaking change. Imho it's better to do the change here rather to ask users to do it.

It is truth that the order of params in `onPress(previousScene, scene, jumpToIndex)  is slightly more intuitive than `onPress(scene, jumpToIndex, previousScene)`, but I don't think this is a reason for making a breaking change. If there is no other reason for doing the breaking change, I'd say we don't do it. Any thoughts?

vonovak replied:
or perhaps a compromise (still breaking, but less), if you want to have a nicer order of params.

change
```
  getOnPress: (
    previousScene: NavigationRoute,
    scene: TabScene
  ) => (
    previousScene: NavigationRoute,
    scene: TabScene,
    jumpToIndex: (index: number) => void
  ) => void,
```

to

```
  getOnPress: (
    scene: TabScene.
    previousScene: NavigationRoute,
  ) => (
    scene: TabScene,
    previousScene: NavigationRoute,
    jumpToIndex: (index: number) => void
  ) => void,
`

pietro909 replied:
I see your point, and to be honest I'm not sure what could be a sensible decision: generally, I prefer a meaningful order for function arguments but the breaking change is not a small one.

Given that this is still a _beta_ version, I'd be ready as a developer to accept breaking changes if needed.

spencercarli replied:
I'm back and forth on this API change. I agree with @pietro909 in that this is beta software, therefore breaking changes will happen. This is also a relatively new feature of the library so I'm not sure how many people are actively using this.

On the flip side of that, we want to encourage confidence and minimize breaking changes and @vonovak's first solution would allow that.

Another question: If we do make a breaking change now, what can we do to ensure we don't have another breaking change in the future? Would it be to pass an object with the arguments in there (eliminates the order issue)?

pietro909 replied:
I find the idea of an object more scalable as a pattern, like:

    ({
        scene: TabScene,
        previousScene: NavigationRoute,
        jumpToIndex: (index: number)
    }) => void

This one offers two advantages:
1. order doesn't matter anymore
2. can add parameters without breaking the API

On the other hand, it makes less clear what is being passed if the consumer is not using Flow (or Typescript). But I think that this is the reason why we have docs :-)

spencercarli replied:
Exactly - docs are there and we can give added flexibility in the future. @vonovak @kelset any hesitations on that API change? Still breaking but "better" breaking? 😄

kelset replied:
⛵️

pietro909 replied:
Do we have any news? Can I proceed with the object instead of single arguments?

kelset replied:
Let's ping @vonovak ;) Btw I think you still need to remove the comment at line 140 commented by him a while back

vonovak replied:
Yes, I was also thinking about that (using a single object argument), I think its okay to do it that way, so let's go ahead with that.

pietro909 commented on src/views/TabView/TabBarTop.js:
```diff
@@ -131,6 +137,10 @@ export default class TabBarTop extends React.PureComponent<Props> {
     // TODO: Define full proptypes
     const props: any = this.props;
 
+    // TODO: get navigtion from props
```
sorry, I forgot it!

pietro909 commented:
I've updated both the sources and the docs. Thanks everybody for the insights :-)

kelset commented:
@vonovak  @spencercarli  I think we are ready to ⛵️

spencercarli commented:
🎉
